Use importmulti timestamp when importing watch only keys (on top of #9682)

ryanofsky opened the pull request:
(This is based on #9682 to avoid merge conflicts in importmulti.py. Only the two commits "Dedup nTimeFirstKey update logic" and "Use importmulti timestamp when importing watch only keys" pertain to this PR. The other three commits belong to #9682).

This fixes issue #9034.

When importing a watch-only address over importmulti with a specific timestamp,
the wallet's nTimeFirstKey is currently set to 1. After this change, the
provided timestamp will be used and stored as metadata associated with
watch-only key. This can improve wallet performance because it can avoid the
need to scan the entire blockchain for watch only addresses when timestamps are
provided.

ryanofsky committed:
Dedup nTimeFirstKey update logic

Also make nTimeFirstKey member variable private.

This is just a cleanup change, it doesn't change behavior in any significant
way.

ryanofsky committed:
Use importmulti timestamp when importing watch only keys

When importing a watch-only address over importmulti with a specific timestamp,
the wallet's nTimeFirstKey is currently set to 1. After this change, the
provided timestamp will be used and stored as metadata associated with
watch-only key. This can improve wallet performance because it can avoid the
need to scan the entire blockchain for watch only addresses when timestamps are
provided.

Also adds timestamp to validateaddress return value (needed for tests).

Fixes #9034.

fanquake commented:
The fundrawtransaction test failed on two linux builds, same error for both:
```
.........
fundrawtransaction.py:
Initializing test directory /tmp/testz6xi8e_p/637
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
Mining blocks...
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
Stopping nodes
Cleaning up
Tests successful
stderr:
Warning: Error reading wallet.dat! All keys read correctly, but transaction data or address book entries might be missing or incorrect.
Pass: False, Duration: 41 s
```

ryanofsky commented:
Fixed fundrawtransaction test, and added importmulti tests.

jonasschnelli commented:
Concept ACK.
Plans to test this soon.

gmaxwell commented:
utACK.

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -222,6 +219,17 @@ bool CWallet::LoadCryptedKey(const CPubKey &vchPubKey, const std::vector<unsigne
     return CCryptoKeyStore::AddCryptedKey(vchPubKey, vchCryptedSecret);
 }
 
+void CWallet::UpdateTimeFirstKey(int nCreateTime)
```
shouldn't this be `int64_t nCreateTime`.

ryanofsky replied:
Good catch, fixed in 6891d5e1f24e74561aeba45c4752756a760c281d.

jonasschnelli commented on src/wallet/walletdb.cpp:
```diff
@@ -454,20 +457,27 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
         {
-            CPubKey vchPubKey;
-            ssKey >> vchPubKey;
+            uint160 keyId;
```
Is there a reason for not using `CKeyID` here?

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -3247,7 +3262,7 @@ void CWallet::GetKeyBirthTimes(std::map<CKeyID, int64_t> &mapKeyBirth) const {
 
     // get birth times for keys with metadata
-    for (std::map<CKeyID, CKeyMetadata>::const_iterator it = mapKeyMetadata.begin(); it != mapKeyMetadata.end(); it++)
+    for (std::map<uint160, CKeyMetadata>::const_iterator it = mapKeyMetadata.begin(); it != mapKeyMetadata.end(); it++)
```
I don't see the reason for changing this from `CKeyID` to `uint160`?

morcos commented:
utACK 6891d5e

jtimon commented:
Concept ACK

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -251,11 +259,18 @@ bool CWallet::AddWatchOnly(const CScript &dest)
 {
```
I'd significantly prefer to not expose this version, making it explicit to callers that the key's nCreateTime is being set to 0 (which, after this PR, implies the wallet's nTimeFirstKey is set to 1).

TheBlueMatt replied:
The second two calls to AddWatchOnly in processImport probably could take the timestamp argument.

TheBlueMatt replied:
Grr, meant for this to be on AddWatchOnly at line https://github.com/bitcoin/bitcoin/pull/9108/files#diff-b2bb174788c7409b671c46ccc86034bdL251

TheBlueMatt commented on src/wallet/walletdb.cpp:
```diff
@@ -116,15 +116,20 @@ bool CWalletDB::WriteCScript(const uint160& hash, const CScript& redeemScript)
 bool CWalletDB::EraseWatchOnly(const CScript &dest)
 {
     nWalletDBUpdated++;
+    if (!Erase(std::make_pair(std::string("watchmeta"), *(const CScriptBase*)(&dest))))
```
I think we need to ignore this error - otherwise if a watchonly has been added in a previous version we will fail to erase here (as it might have no watchmeta)?

TheBlueMatt commented on src/wallet/walletdb.cpp:
```diff
@@ -116,15 +116,20 @@ bool CWalletDB::WriteCScript(const uint160& hash, const CScript& redeemScript)
 {
     nWalletDBUpdated++;
+    if (!Write(std::make_pair(std::string("watchmeta"), *(const CScriptBase*)(&dest)),
+               keyMeta, false))
```
Maybe do overwrite here? Seems strange that we update in-wallet-state without updating on-disk state (and its probably a supported use-case to re-add something to wallet to set its date differently).

jtimon commented on src/wallet/walletdb.cpp:
```diff
@@ -463,11 +463,6 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
             wss.nKeyMeta++;
 
             pwallet->LoadKeyMetadata(vchPubKey, keyMeta);
-
```
Missing this here?
```
pwallet->UpdateTimeFirstKey(keyMeta.nCreateTime);
```

jtimon commented:
utACK individual commits: 99c1aa169592a71c24e7a2475c50ad7598eafb65 6891d5e1f24e74561aeba45c4752756a760c281d

Can we squash already?

ryanofsky commented:
Squashed 6891d5e1f24e74561aeba45c4752756a760c281d -> 89011a3032f635b6536a476bfa48502c5dcc7a65, and will look into Matt's comments.

ryanofsky commented:
Squashed 1b9969682e59e082855179e4c50fb1dd38227d67 -> 3d6dbed2c3132d53650b86f09d42e339149d3210 ([watchtime.3](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.3) -> [watchtime.4](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.4))

ryanofsky commented:
Latest changes conflicted (trivially) with #9607 and #9121, so rebased 3d6dbed2c3132d53650b86f09d42e339149d3210 -> b78ec3cc3dbf5e1fc2dc105c609574101e462ae8 ([watchtime.4](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.4) -> [watchtime.5](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.5))

ryanofsky committed:
Require timestamps for importmulti keys

Additionally, accept a "now" timestamp, to allow avoiding rescans for keys
which are known never to have been used.

Note that the behavior when "now" is specified is slightly different than the
previous behavior when no timestamp was specified at all. Previously, when no
timestamp was specified, it would avoid rescanning during the importmulti call,
but set the key's nCreateTime value to 1, which would not prevent future block
reads in later ScanForWalletTransactions calls. With this change, passing a
"now" timestamp will set the key's nCreateTime to the current block time
instead of 1.

Fixes #9491

TheBlueMatt commented on src/wallet/wallet.h:
```diff
@@ -603,6 +603,18 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
+     * Because this is an inherited virtual method, it is accessible despite
+     * being marked private, but it is marked private anyway to encourage use of
+     * the other AddWatchOnly which accepts a timestamp and sets nTimeFirstKey
+     * more intelligently for more efficient rescans.
```
Ugh, I totally missed that its an override, thanks for catching that. Can you further document here that this method implies setting the first key time to 1?

ryanofsky replied:
Added in 5dbd05d8f7c4fc86d1296a200a00a834b61fb5c3

TheBlueMatt commented:
utACK b78ec3cc3dbf5e1fc2dc105c609574101e462ae8 +/- wanting another comment, Didn't look too closely at tests.

ryanofsky reviewed (commented):
Added comment in 5dbd05d8f7c4fc86d1296a200a00a834b61fb5c3.

Squashed 5dbd05d8f7c4fc86d1296a200a00a834b61fb5c3 -> b8fca20399c3528e5c293509629907f6874575e3 ([watchtime.6](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.6) -> [watchtime.7](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.7))

ryanofsky committed:
Add test to check new importmulti "now" value

Easiest way to test this was to expose the timestamp via the validateaddress
RPC (which was already looking up and returning key metadata).

ryanofsky commented:
Rebased b8fca20399c3528e5c293509629907f6874575e3 -> be820d8ad761b87ca61427ce8b3005b0f6db0d06 ([watchtime.7](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.7) -> [watchtime.8](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.8)) to fix conflict with #9227.

TheBlueMatt commented:
be820d8ad761b87ca61427ce8b3005b0f6db0d06 appears to be equivalent to my previous utACK (b78ec3cc3dbf5e1fc2dc105c609574101e462ae8) plus a fixup for the comments I requested.

ryanofsky committed:
Use MTP for importmulti "now" timestamps

laanwj commented:
@jonasschnelli Looks like your suggestions have been carried out by @ryanofsky  and the "changes requested" can be cancelled?

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -222,6 +219,17 @@ bool CWallet::LoadCryptedKey(const CPubKey &vchPubKey, const std::vector<unsigne
+    AssertLockHeld(cs_wallet);
+    // Cannot determine birthday information, so set the wallet birthday to the
+    // beginning of time.
+    if (nCreateTime <= 1)
```
Maybe I'm misreading the code, but the comment `Cannot determine birthday information` seems to refer to the if case below it being true; I would prefer comments being inside the block in this case.  I would probably also include brackets after doing so as the comment makes the single-line conditional state less obvious.

ryanofsky replied:
Fixed in f95843a294dad786117bd40801c8bf40afcb8784

kallewoof reviewed (approved):
Slight/weak ACK (some testing but not a lot)

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -576,15 +575,17 @@ UniValue dumpwallet(const JSONRPCRequest& request)
     std::vector<std::pair<int64_t, CKeyID> > vKeyBirth;
-    for (std::map<CKeyID, int64_t>::const_iterator it = mapKeyBirth.begin(); it != mapKeyBirth.end(); it++) {
-        vKeyBirth.push_back(std::make_pair(it->second, it->first));
+    for (auto it = mapKeyBirth.begin(); it != mapKeyBirth.end(); it++) {
```
Fixed in 6a56f6d77a4e0679ab88b354dea7143ec8539c4e

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -576,15 +575,17 @@ UniValue dumpwallet(const JSONRPCRequest& request)
-    for (std::map<CKeyID, int64_t>::const_iterator it = mapKeyBirth.begin(); it != mapKeyBirth.end(); it++) {
-        vKeyBirth.push_back(std::make_pair(it->second, it->first));
+    for (auto it = mapKeyBirth.begin(); it != mapKeyBirth.end(); it++) {
+        if (const CKeyID* keyID = boost::get<CKeyID>(&it->first)) {
```
Fixed in 6a56f6d77a4e0679ab88b354dea7143ec8539c4e

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3416,12 +3431,12 @@ class CAffectedKeysVisitor : public boost::static_visitor<void> {
 
     // get birth times for keys with metadata
-    for (std::map<CKeyID, CKeyMetadata>::const_iterator it = mapKeyMetadata.begin(); it != mapKeyMetadata.end(); it++)
+    for (auto it = mapKeyMetadata.begin(); it != mapKeyMetadata.end(); it++)
```
Fixed in 6a56f6d77a4e0679ab88b354dea7143ec8539c4e

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -748,7 +763,7 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
 
     //! Adds a watch-only address to the store, and saves it to disk.
-    bool AddWatchOnly(const CScript &dest);
+    bool AddWatchOnly(const CScript &dest, int64_t nCreateTime);
```
Nit 2 fixed in b3ba65488d575899c6df2a98f9dc6447c93ae22d

ryanofsky commented:
Squashed b3ba65488d575899c6df2a98f9dc6447c93ae22d -> f35ce5172bb9c6bec6bb636ea349104b559d36de ([watchtime.10](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.10) -> [watchtime.11](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.11))

jonasschnelli commented:
UtACK f35ce5172bb9c6bec6bb636ea349104b559d36de

ryanofsky commented:
Rebased on top of #9682 to avoid merge conflicts in the importmulti.py test. No other changes.

f35ce5172bb9c6bec6bb636ea349104b559d36de -> a80f98b1c7a49432dc53d18d0fb51ac334de96be ([watchtime.11](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.11) -> [watchtime.12](https://github.com/ryanofsky/bitcoin/commits/pr/watchtime.12))

jonasschnelli commented:
Not sure if this is a nit or the intended behaviour:
If I re-import a watch only address (same address I have already imported), I get no notice that it was already present. This may be okay.

Another option would be to reject the import with "already exists". But the current overwrite-behaviour allows to re-set the timestamp which can be handy.

jonasschnelli commented:
Correction of my comment above.
If you re-import a watch-only, it does not set the new timestamp but still response with "success". I think we should do this (fail when reimport a watch-only)... 
```diff
diff --git a/src/wallet/rpcdump.cpp b/src/wallet/rpcdump.cpp
index 0a32259..401614c 100644
--- a/src/wallet/rpcdump.cpp
+++ b/src/wallet/rpcdump.cpp
@@ -922,9 +922,12 @@ UniValue ProcessImport(const UniValue& data, const int64_t timestamp)
                     throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
                 }
 
-                pwalletMain->MarkDirty();
+                if (pwalletMain->HaveWatchOnly(script)) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains this address or script");
+                }
 
-                if (!pwalletMain->HaveWatchOnly(script) && !pwalletMain->AddWatchOnly(script, timestamp)) {
+                pwalletMain->MarkDirty();
+                if (!pwalletMain->AddWatchOnly(script, timestamp)) {
                     throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
                 }
 

```

An alternative would be to overwrite the timestamp with the new one.

ryanofsky commented:
Ok, I can add logic to error on attempts to import watch only keys that have already imported. Note that your fix above is incomplete because it only covers `importmulti`, not `importaddress`, `importpubkey`, etc.

I'll do it in a different PR though, unless you think it needs to be here, because this is longstanding behavior not really related to saving timestamps.

jonasschnelli commented:
Tested ACK of a58370e6a2d4dce50eefbcab5bde9f14facef8fc a80f98b1c7a49432dc53d18d0fb51ac334de96be on top of master (e87ce95fbdc6ca6ef822c978d98b2acba5948ee1).
I agree with @ryanofsky that the overwrite-/already-exists-behavior can be fixed later.

morcos commented:
re-utACK a80f98b

sipa commented:
utACK a80f98b1c7a49432dc53d18d0fb51ac334de96be
